trunk-recorder/main.cc

```diff
@@ -954,6 +954,75 @@ void handle_call(TrunkMessage message, System *sys) {
 
 }
 
+void clear_stale_moto_patches(System *sys){
+  std::map<int,std::time_t> patch_map;
+  std::vector<std::map<int,std::time_t>> current_patches = sys->get_talkgroup_patches();
+  std::vector<std::map<int,std::time_t>> updated_patches;
+  //for each patch map in the vector:
+  BOOST_FOREACH (auto& patch, current_patches) {
+    //for each TGID/element in the patch map:
+    patch_map.clear();
+    BOOST_FOREACH(auto& patch_element, patch){
+      if (std::time(nullptr) - patch_element.second < 3){  //3 second hard coded timeout for now
+        //only add back into our active patch map if the timestamp associated with this TGID is less than three seconds old
+        patch_map[patch_element.first] = patch_element.second; 
+      }
+    }
+    if(!patch_map.empty()){
+      updated_patches.push_back(patch_map);
+    }
+  }
+  BOOST_LOG_TRIVIAL(info) << "Found " << updated_patches.size() << " active Moto patches";
+  BOOST_FOREACH (auto& patch, updated_patches) {
+    string printstring;
+    BOOST_FOREACH(auto& patch_element, patch){
+      printstring+=" ";
+      printstring+= std::to_string(patch_element.first);
+    }
+    BOOST_LOG_TRIVIAL(info) << "Active Moto Patch of TGIDs" << printstring;
+  }
+  sys->set_talkgroup_patches(updated_patches);
+}
+
+void update_moto_patches(TrunkMessage message, System *sys){
+  std::map<int,std::time_t> patch_map;
+  std::time_t update_time = std::time(nullptr);
+  std::vector<std::map<int,std::time_t>> current_patches = sys->get_talkgroup_patches();
+  std::vector<std::map<int,std::time_t>> updated_patches;
+  bool new_flag = true;
+  //for each patch map in the vector:
+  BOOST_FOREACH (auto& patch, current_patches) {
+    //for each TGID/element in the patch map:
+    patch_map.clear();
+    BOOST_FOREACH(auto& patch_element, patch){
+      patch_map[patch_element.first] = patch_element.second; 
+      //if the TGID == one of the TGIDs from our Message:
+      if (patch_element.first == (int)message.patch_data.sg || patch_element.first == int(message.patch_data.ga1) || patch_element.first == int(message.patch_data.ga2) || patch_element.first == int(message.patch_data.ga3)){
+        new_flag = false;
+        //BOOST_LOG_TRIVIAL(debug) << "Found existing patch to update";
+        patch_map[(int)message.patch_data.sg] = update_time;
+        patch_map[(int)message.patch_data.ga1] = update_time;
+        patch_map[(int)message.patch_data.ga2] = update_time;
+        patch_map[(int)message.patch_data.ga3] = update_time;
+      }
+    }
+    if(!patch_map.empty()){
+      updated_patches.push_back(patch_map);
```

Comment:
Yeah I'm just duplicating/updating patches that are already there and then adding in any that are new.  I'm creating an entirely new vector of patches during iteration that's based on the existing patch vector + the latest message rather than modifying the patch existing vector while iterating.  There may be a better/more elegant way to do this.

Reply:
This is the only part of the approach that is giving me a little pause. Each new message results in a complete rebuild. This wouldn't be a problem if there are only a few patches. It would be better to update or add in place. I might see if I can restructure this a little.

Reply:
Thinking about it more - this may all be optimized by the compiler. If things are slow we could move this section into System and have it work directly on the Vector/Map.

Reply:
I'll try to give it some more thought tonight.  One additional thing to consider is the possibility of patches with more than 4 TGIDs.  Each Message only has space for four (sg, ga1, ga2, ga3).  I haven't seen patches with more than four on my local system yet, but my assumption is that multiple messages get sent, each with the same sg but with different/additional ga1,ga2,ga3 values in each message.  So the full set of patch data may span multiple messages. 

I suppose if we used a map of maps with the supergroup as the key for the outer map, that may make modification in place a bit easier? It may also save a bit on iterations...when a call becomes active it can quickly check for the supergroup key in the outer map and if it's found and not have to loop through every TGID of every patch.  

Then for expiry,check each inner TGID and if the inner map ends up empty, also kill the outer map.  

Agree it probably makes sense to move the update & stale checking code into System so that they can modify directly and just call those functions from main.

```
 std::map<int,std::time_t> individual_patch;
std::map<int, std::map<int,std::time_t>> all_patches_map
individual_patch[tgid] = timestamp;
all_patches_map[sg] = patch;
```